packages/ai/core/tool/mcp.ts

```diff
@@ -0,0 +1,227 @@
+import { ToolSet } from './../generate-text/tool-set';
+import { inferParameters, Tool, tool } from './tool';
+import { z } from 'zod';
+import { IOType } from 'node:child_process';
+import { Stream } from 'node:stream';
+import { SimpleMcpClient } from './mcp/client';
+import {
+  CallToolRequest,
+  CallToolResult,
+  CallToolResultSchema,
+  CompatibilityCallToolResultSchema,
+  Implementation,
+  ListToolsRequest,
+  ListToolsResult,
+  RequestOptions,
+} from './mcp/types';
+
+interface McpStdioServerConfig {
+  command: string;
+  args?: string[];
+  env?: Record<string, string>;
+  stderr?: IOType | Stream | number;
+  cwd?: string;
+  type: 'stdio';
+}
+interface McpSSEServerConfig {
+  type: 'sse';
+  url: string;
+  options?: {
+    // authProvider?: OAuthClientProvider;
+    eventSourceInit?: EventSourceInit;
+    requestInit?: RequestInit;
+  };
+}
+
+const DEFAULT_CLIENT_CONFIG = {
+  name: 'ai-sdk-mcp-client',
+  version: '1.0.0',
+  connectionTimeoutMs: 6000,
+  requestTimeoutMs: 3000,
+};
+
+/**
+ * An interface that defines the minimum required MCP client properties.
+ * Used when endusers provide their own MCP Client for more granular control.
+ */
+interface McpClientInterface {
+  listTools(
+    params?: ListToolsRequest['params'],
+    options?: RequestOptions,
+  ): Promise<ListToolsResult>;
+  callTool(
+    params: CallToolRequest['params'],
+    resultSchema?:
+      | typeof CallToolResultSchema
+      | typeof CompatibilityCallToolResultSchema,
+    options?: RequestOptions,
+  ): Promise<any>;
+}
+
+interface AdapterConfig {
+  server: McpStdioServerConfig | McpSSEServerConfig;
+  customClient?: McpClientInterface;
+  clientConfig?: Implementation;
+  toolCallOptions?: RequestOptions;
+}
+
+interface AdapterReturnType<TOOL_SCHEMAS extends ToolSchemas = {}> {
+  toolSet: McpToolSet<TOOL_SCHEMAS>;
+  cleanup: () => Promise<void>;
+}
+
+type ToolSchemas = Record<string, { input: z.ZodTypeAny | Schema<any> }>;
+
+type McpToolSet<TOOL_SCHEMAS extends ToolSchemas = {}> = ToolSet & {
+  [K in keyof TOOL_SCHEMAS]: Tool<
+    inferParameters<TOOL_SCHEMAS[K]['input']>,
+    CallToolResult
+  > & {
+    execute: (
+      args: inferParameters<TOOL_SCHEMAS[K]['input']>,
+      // options?
+    ) => Promise<CallToolResult>;
+  };
+} & {
+  [key: string]: Tool<any, CallToolResult> & {
+    execute: (
+      args: any,
+      // options?
+    ) => Promise<CallToolResult>;
+  };
+};
+
+export async function createMcpTools<TOOL_SCHEMAS extends ToolSchemas = {}>(
+  config: AdapterConfig,
+  toolSchemas: TOOL_SCHEMAS,
+): Promise<AdapterReturnType<TOOL_SCHEMAS>> {
+  const tools: Record<string, Tool> = {};
+  const { client, cleanup } = await setupClient(config);
+
+  try {
+    const listToolsResult = await client.listTools();
+
+    for (const { name, description, inputSchema } of listToolsResult.tools) {
+      const parameters = toolSchemas[name]
+        ? toolSchemas[name].input
+        : jsonSchema(inputSchema as JSONSchema7);
```

Comment:
hm this might lead to errors where toolSchemas is wrong

Reply:
very similar to the dynamic nature of fetching all available tools, regardless of whether the user explicitly defined in toolSchemas; would love to chat about this further